feat(docs-infra): generate Angular CLI command reference

petebacondarwin opened the pull request:
This is a first cut - for demoing and discussion - of the rendering of the Angular CLI command reference in the angular.io docs website.

@Brocco  - I couldn't add you as an assignee. But I think you should be involved 🤗

petebacondarwin commented on aio/tools/transforms/cli-docs-package/clone-cli-repo.js:
```diff
@@ -0,0 +1,11 @@
+const {resolve}  = require('canonical-path');
+
+const REPO = 'https://github.com/hansl/angular-cli';
+const BRANCH_OR_TAG = 'commands-json-schema';
```
Note that `REPO` is currently pointing to @hansl's fork because the JSON files have not yet landed in the main repo. Eventually this will change to `https://github.com/angular/angular-cli`.

Also, `BRANCH_OR_TAG` is currently pointing to the branch for the PR containing the JSON files, but eventually this will need to be set to the tagged commit that we want to generate CLI docs for. E.g. a release tag.

petebacondarwin commented on aio/tools/transforms/cli-docs-package/index.js:
```diff
@@ -0,0 +1,54 @@
+  const COMMANDS_FILE = readFileSync(COMMANDS_PATH, 'utf8');
+  const commands = json.parse(COMMANDS_FILE);
+  return Object.keys(commands).map(name => resolve(basePath, commands[name]));
+}
```
We read the paths to each of the command JSON files from the `commands.json` file that must be in a reliable place.

petebacondarwin commented on aio/tools/transforms/cli-docs-package/processors/filterHiddenCommands.spec.js:
```diff
@@ -0,0 +1,23 @@
+    const processor = processorFactory();
+    expect(processor.$runBefore).toEqual(['processCliContainerDoc']);
+  });
+});
```
TODO: add actual tests of this processors functionality.

petebacondarwin commented on aio/tools/transforms/cli-docs-package/processors/processCliContainerDoc.spec.js:
```diff
@@ -0,0 +1,23 @@
+    const processor = processorFactory();
+    expect(processor.$runBefore).toEqual(['rendering-docs']);
+  });
+});
```
TODO: add actual tests of this processors functionality.

petebacondarwin replied:
Done

petebacondarwin commented on aio/tools/transforms/cli-docs-package/readers/cli-command.js:
```diff
@@ -0,0 +1,72 @@
+  // Ensure the argument array is in the correct order
+  arguments.sort((a,b) => a.$default.index - b.$default.index);
+  return { arguments, options };
+}
```
We may need to augment this processing to capture additional information about the constraints on the properties.

petebacondarwin commented on aio/tools/transforms/cli-docs-package/.gitignore:
```diff
@@ -0,0 +1 @@
+tmp
```
We clone the Angular CLI repo into this `tmp` folder, which will be ignored by git.

petebacondarwin commented on aio/tools/transforms/templates/cli/cli.template.html:
```diff
@@ -0,0 +1,25 @@
+  </tr>
+{% endfor %}
+</tbody>
+</table>
```
The name of this template matches the `id` of the cli overview document. This has a higher priority in the dgeni template finder than the docType template (content.template.html) which is normally used for docs such as guides.

mary-poppins commented:
You can preview ca3b449 at https://pr25363-ca3b449.ngbuilds.io/.

JoostK commented:
This is very cool!

@petebacondarwin I am missing a lot of CLI options for each task, e.g. [`serve` only shows `configuration` and `prod`](https://pr25363-ca3b449.ngbuilds.io/cli/serve). Is that something that could be improved?

hansl commented:
@petebacondarwin Would it be useful to add hints for default schematics and default build system? So that you can add those as a "if you didn't change your angular.json, it will run the default schematics which has the following options"?

Maybe useful as a talking point in our next meeting.

alexeagle commented:
Update from meeting yesterday: the static files checked into the CLI repo aren't sufficient to generate help - you need to dynamically run the schematics code to stitch in the options.
We'll run `ng help` as part of the CI build for the CLI, and publish a JSON-formatted output for all the help commands in our package snapshots.
Then the dgeni process for angular.io can fetch our snapshot repo and will have the complete data there.

hansl commented:
TOOL-166 and TOOL-167 have been created to track the work from the CLI part (JSON help, and snapshots).

gkalpak commented on aio/content/cli.md:
````diff
@@ -0,0 +1,83 @@
+Invoke the tool on the command line through the ng executable. Online help is available on the command line:
+
+```
+> ng help		  	        Lists commands with short descriptions
````
Nit: This is not aligned in the resulting HTML.

petebacondarwin replied:
Cool. This will be a rewrite when we get the JSON help files from the CLI.

jenniferfell commented:
Quickly making sure to capture the nav things that we discussed when this first was available. 

Under CLI in the navigation, it would be nice to have children:

CLI command reference
- Introduction (in-page title "Introduction to the CLI")
- add
- build
- config
- etc.

jenniferfell commented:
@petebacondarwin We have a draft of the desired output, including suboptions/commands of schematics commands. It's been in review since June. I'll repost the link in our slack channel.

mary-poppins commented:
You can preview 32b1d98 at https://pr25363-32b1d98.ngbuilds.io/.

petebacondarwin commented:
OK, so check this out. I have updated the code to pull in the JSON help from the Angular CLI.
The current rendered output is not yet right but you can get the basic idea in the preview: https://pr25363-6f4544b.ngbuilds.io/cli/add.

mary-poppins commented:
You can preview bee02b4 at https://pr25363-bee02b4.ngbuilds.io/.

petebacondarwin commented:
TODO:

* remove alias columns
* add positional stuff to schematics
* flow the string options on the same line
* remove `true|false` and replace with `--flag` or `--no-flag` depending upon default.
* add a syntax line to schematics

petebacondarwin committed:
build(docs-infra): add option to run only the doc-gen

This can save time when iterating by not
regenerating the zips and embedded examples.

petebacondarwin committed:
feat(docs-infra): generate Angular CLI command reference

mary-poppins commented:
You can preview 16b9517 at https://pr25363-16b9517.ngbuilds.io/.

jenniferfell commented:
Quick thought: Perhaps we should alphabetize the schematics in commands such as ng generate?

hansl commented on aio/tools/transforms/cli-docs-package/processors/processCliCommands.js:
```diff
@@ -0,0 +1,54 @@
+  container.namedOptions = [];
+  container.options = options.filter(option => !option.hidden);
+  container.options.forEach(option => {
+    if (!option.hidden) {
```
This is redundant with the above `filter(...)` on line 39.

petebacondarwin commented on aio/content/cli/tmp/package.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "dependencies": {
+    "@angular/cli": "https://github.com/angular/cli-builds#master"
```
The trick is that we should change the branch here to match the branch that we are releasing angular.io from.

petebacondarwin committed:
fixup! feat(docs-infra): generate Angular CLI command reference

mary-poppins commented:
You can preview 9a5a51a at https://pr25363-9a5a51a.ngbuilds.io/.

jenniferfell commented:
In current preview, syntax for aliases is not showing up. Instead full syntax shows up twice. 

Also from Judy in slack: 
Neither syntax statement shows the argument, which I would expect to be the arg name in italic: `ng generate *schematic* [*options*]`
...or, since there are both command options as subcommand options, perhaps `ng generate *schematic* [*command_options*][*subcommand_options*]`
The italics distinguish placeholders (as opposed to constant values like option names), so I think they are especially important the formal syntax statement. Other commands do have named arguments that would show up here - `name=*app_name*` (edited)
I think it works well to have the sub-command options on the same page.

jenniferfell commented:
Can we alphabetize the schematics in commands like ng generate?

petebacondarwin committed:
fixup! feat(docs-infra): generate Angular CLI command reference

mary-poppins commented:
You can preview 9573c8a at https://pr25363-9573c8a.ngbuilds.io/.

petebacondarwin committed:
fixup! feat(docs-infra): generate Angular CLI command reference

petebacondarwin commented:
@jenniferfell - the subcommands are now alphabetised. I have also fixed the bugs that @jbogarthyde noted. Also I made a few changes in line with your interview with @johnpapa. Notably I have implemented all of the rendering options for options [sic]. You can choose which we through away.

mary-poppins commented:
You can preview 2a83ac8 at https://pr25363-2a83ac8.ngbuilds.io/.

jenniferfell reviewed (approved):
Nav looks good! Command reference pages look good for RC. We can continue our group discussions/refinements during RC period. Thanks!!

gkalpak commented on aio/package.json:
```diff
@@ -41,7 +42,8 @@
     "predocs": "yarn generate-stackblitz && yarn generate-zips",
-    "docs": "dgeni ./tools/transforms/angular.io-package",
+    "docs": "yarn docs-only",
+    "docs-only": "dgeni ./tools/transforms/angular.io-package",
```
Why this change?

petebacondarwin replied:
Often I didn't want the overhead of the Stackblitz and Zip file creation just to iterate on my changes to the doc-gen.

gkalpak commented on aio/tools/transforms/cli-docs-package/extract-cli-commands.js:
```diff
@@ -0,0 +1,7 @@
+
+shelljs.cd(resolve(CONTENTS_PATH, 'cli-src'));
+shelljs.exec('git clean -Xfd');
+shelljs.exec('yarn install');
```
This script doesn't seem to extract anything 🤔
Consider renaming to something more descriptive.

petebacondarwin replied:
It extracts the CLI library from npm :-P
Suggestion for a better name?

gkalpak reviewed (approved):
LGTM

For the record, I will say that I am not a big fun of adding the `CLI COMMANDS` as a top-level menu in the sidenav. I would rather have a `CLI` submenu (above `API`).

petebacondarwin commented:
I believe that the navigation side of things is still under discussion.

petebacondarwin commented:
We will need to create a new PR for patch, since the `aio/content/cli-src/package.json` needs changing to map to the correct version of Angular CLI for those docs.

petebacondarwin commented:
OK, so this is only for master, which is why we can experiment with the nav a bit.
